server/services/venueservices.go

```diff
@@ -0,0 +1,152 @@
+package services
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"server/models"
+
+	"github.com/gin-gonic/gin"
+)
+
+// GET /home
+// get all venues
+func GetVenues(c *gin.Context) {
+	// get all the venues
+	var SearchPage []models.SearchPage
+	query1 := "SELECT * FROM public.Venues;"
+	if err := DB.Raw(query1).Scan(&SearchPage).Error; err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	// get all facilities in the venue and put into homepage struct
+	var venuesandfacilities []models.VenuesAndFacilities
+
+	for i, s := range SearchPage {
+		SearchPage[i].Facilitiesdict = make(map[string]int)
+		tempQuery := "SELECT * FROM public.Venuefacilities " +
+			"JOIN public.Facilities ON public.Venuefacilities.facilityid = public.facilities.id " +
+			"WHERE venueid = ?;"
+		if err := DB.Raw(tempQuery, s.ID).Scan(&venuesandfacilities).Error; err != nil {
+			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+			return
+		}
+
+		// make dict that contains all equipment and its quantity
+		for _, y := range venuesandfacilities {
+			SearchPage[i].Facilitiesdict[y.Facilityname] = y.Quantity
+		}
+	}
+
+	venueArrayDict := make([]map[string]string, 0)
+	for _, s := range SearchPage {
+		temp := make(map[string]string)
+		temp["Venuename"] = s.Venuename
+		venueArrayDict = append(venueArrayDict, temp)
+	}
+
+	c.JSON(http.StatusOK, gin.H{"data": venueArrayDict})
+	log.Println("Return successful!")
+}
+
+// POST /search
+// search for venues based on filters applied
+func SearchVenues(c *gin.Context) {
+	// first get array of facilities
+	var searchInput models.SearchInput
+	if err := c.ShouldBindJSON(&searchInput); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"success": false, "message": "Check input fields. " + err.Error()})
+		fmt.Println("Error in getting search input. " + err.Error() + "\n")
+		return
+	}
+
+	// retrieve array of facilityID by querying using facilityname
+	var facility models.Facilities
+	facilityIDArr := make([]int, 0)
+	for _, s := range searchInput.Equipment {
+		facilityQuery := "SELECT id FROM facilities WHERE name = ?"
+		if err := DB.Raw(facilityQuery, s).Scan(&facility).Error; err != nil {
+			c.JSON(http.StatusBadRequest, gin.H{"error": err.Error(), "message": "Error in querying for facilityID. "})
+			fmt.Println("Check facilityQuery. " + err.Error() + "\n")
+			return
+		}
+		facilityIDArr = append(facilityIDArr, facility.ID)
+	}
+
+	// get all the venueid, from venuefacilities table, with equipment filter applied
+	var venuefacility []models.Venuefacilities
+	venueFacilityQuery := "SELECT venueid FROM public.Venuefacilities ORDER BY venueid"
+	for _, s := range facilityIDArr {
+		venueFacilityQuery += fmt.Sprintf(" INTERSECT SELECT venueid FROM public.Venuefacilities WHERE facilityid = %d", s)
```

Comment:
Consider this query:

Let's say if the `facilityid` required is `1,3.` Then the query should be:
```
SELECT venueid FROM public.Venuefacilities
WHERE facilityid IN (1,3)
GROUP BY venueid
HAVING COUNT(facilityid) = 2
```
Let's say if all facilityids are required. Then the query is simply:
```
SELECT venueid FROM public.Venuefacilities
```